### DISCUSSION TITLE
Long forms

### DISCUSSION THREAD

**Discussion by Pachat** [directus/directus#12267] [Archive]
Thanks to versatility of Directus I could port about 90% of a PHP development with (very) long forms: 50 fields including 8 relational.
Such a form will be filled in many sessions, from different devices and networks.

The problem is to save the input progressively and automatically just as Ctrl+S does.

- Ctrl+S on a tablet?
- Even on a laptop, focused on the information to enter, the user may forget about saving when closing the device.

So what I'm trying to find is to automate saving while the form is fed.
_The PHP + Javascript version is sending the content of an input to the server `onblur `of that input._

Saving each significant part of the form instead of each input is also acceptable.

The splitting of my form is done by an Accordion containing 10 raw groups, all closed at start.
_I like the Accordion for it allows to present a sort of index (with longer names than tabs would allow)
and that can be read at one glance in case the user wants to go directly to some part of the form (better than a Previous/Next organization between small forms)._

A nice place to auto save the group would be on closing it (including closing the form itself).

With Javacript, I could set an event on clicking a group. (Storing it #12266 would be great).
What do you think of this approach?

**Comment by rijkvanzanten** [comment]
The issue with saving the item is that it's "committing" the changes to the database. You won't be able to cancel out anymore for example, or easily undo the changes, as they are already persisted in the DB. Similarly, auto-generated fields like auto-generated IDs / timestamps / etc might get into unexpected states, as the item is technically already created. (For example, if I take 3 days to fill out the form and hit "save", I expect the "date created" field to be today, rather than 3 days ago. Actually saving the item to the DB also means that the incomplete item will already show up in the API responses (as it's now an actual item), and that you can't really have "required" or other validation anymore, as the item needs to be able to save in this incomplete state.

I think we should think through alternative ways to persist the current edit state, without actually committing to saving the item to the database 🤔 (For example local-storage, or keep the editing state in redis before it's being saved, etc)

**Reply by benhaynes** [reply]
As a Core feature, I agree with what Rijk said here — though you could perhaps write your own logic to handle this if it works for your project.

I agree that saving things to local storage is the best option for Core. Then we can literally save as you go... all the way down to each keypress if we wanted to. It's the same as what GitHub does here... if you refresh the page while commenting, your comment is still there, ready to be saved.

**Reply by Pachat** [reply]
Thank you for elaborating on this subject.

- Save locally will not allow the user to continue feeding the form on another device 
_(Beginning at work on a computer, having some precision to put while traveling with his laptop, etc.)._

- Undo changes is not a required feature here. 
Either, it does not matter to have as created date the beginning or the ending of the feeding. 
And there is no required fields (apart the Name). 
_Saving_ is just seeing here as persisting the form as is, allowing to work on it later on another device. 
It's only when the user has a complete form that he hits a special `Submit` button (different from `Save`) where some logic can tell him eventually what is still missing.

I understand that this cannot be implemented on the fly.

So another solution would be going to multi-steps forms.
In that perspective, is there a way to route the `Save` button to another URL than the Collection?

**Reply by rijkvanzanten** [reply]
There's currently no out of the box way to have the form save to a different collection than the one you're currently editing

**Reply by Pachat** [reply]
Thank you for your answer.
It looks like either I'll have:

- to write many extensions: modules, interfaces, etc. which is a lot of work to redo the 90% Directus offers to add the 10% missing
- or build a complete Front end (Next, Nuxt, ...) using Directus only as headless.

**Reply by Pachat** [reply]
@rijkvanzanten To keep things simple
> Similarly, auto-generated fields like auto-generated IDs / timestamps / etc might get into unexpected states, as the item is technically already created. (For example, if I take 3 days to fill out the form and hit "save", I expect the "date created" field to be today, rather than 3 days ago. Actually saving the item to the DB also means that the incomplete item will already show up in the API responses (as it's now an actual item), and that you can't really have "required" or other validation anymore, as the item needs to be able to save in this incomplete state.

We already have  `Save and stay` (Ctrl+S) and it works well: the `Date created` is set first and then the `Date updated` which is OK.

> I think we should think through alternative ways to persist the current edit state, without actually committing to saving the item to the database 🤔 (For example local-storage, or keep the editing state in redis before it's being saved, etc)

We already have the Status: `Draft` enough to differentiate the state while the saved form is not completed.
At completion, the user sets the Status to `published` which tells that the form is ready to be evaluated.

**Reply by rijkvanzanten** [reply]
> @rijkvanzanten To keep things simple
> 
> > Similarly, auto-generated fields like auto-generated IDs / timestamps / etc might get into unexpected states, as the item is technically already created. (For example, if I take 3 days to fill out the form and hit "save", I expect the "date created" field to be today, rather than 3 days ago. Actually saving the item to the DB also means that the incomplete item will already show up in the API responses (as it's now an actual item), and that you can't really have "required" or other validation anymore, as the item needs to be able to save in this incomplete state.
> 
> We already have `Save and stay` (Ctrl+S) and it works well: the `Date created` is set first and then the `Date updated` which is OK.

I don't know if I agree. When you're talking about a long editing process (multiple hours/days), then the date created should be set to the date when the item was completed, not the date the item was first soft-saved IMO.

> > I think we should think through alternative ways to persist the current edit state, without actually committing to saving the item to the database 🤔 (For example local-storage, or keep the editing state in redis before it's being saved, etc)
> 
> We already have the Status: `Draft` enough to differentiate the state while the saved form is not completed. At completion, the user sets the Status to `published` which tells that the form is ready to be evaluated.

I disagree with this. The status field is just an arbitrary data point, so I wouldn't want to enable/disable all validation based on a single field value. The end goal is to make sure that staged edits don't get lost between sessions right? I believe there's better ways to achieve that without saving the item ahead of time 🤔
